src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -590,22 +582,52 @@ public Object evalBindingCached(
             final MaterializedFrame parentFrame = BindingNodes.getTopFrame(binding);
+            return eval(cachedRootNode, cachedCallTarget, callNode, parentFrame);
+        }
+
+        private Object eval(final RootNodeWrapper rootNode, final CallTarget callTarget, final DirectCallNode callNode, final MaterializedFrame parentFrame) {
             final Object bindingSelf = RubyArguments.getSelf(parentFrame);
 
             final InternalMethod method = new InternalMethod(
                     getContext(),
-                    cachedRootNode.getRootNode().getSharedMethodInfo(),
+                    rootNode.getRootNode().getSharedMethodInfo(),
                     RubyArguments.getMethod(parentFrame).getLexicalScope(),
-                    cachedRootNode.getRootNode().getSharedMethodInfo().getName(),
+                    rootNode.getRootNode().getSharedMethodInfo().getName(),
                     RubyArguments.getMethod(parentFrame).getDeclaringModule(),
                     Visibility.PUBLIC,
-                    cachedCallTarget);
+                    callTarget);
 
             return callNode.call(RubyArguments.pack(parentFrame, null, method, RubyArguments.getDeclarationContext(parentFrame), null, bindingSelf, null, new Object[]{}));
         }
 
         @Specialization(guards = {
                 "isRubyString(source)",
+                "equalNode.execute(rope(source), cachedSource)",
+                "isRubyBinding(binding)",
+                "!assignsNoNewVariables(cachedRootNode)",
+                "assignsNoNewVariables(rootNodeToEval)",
+                "bindingDescriptor == getBindingDescriptor(binding)"
+        }, limit = "getCacheLimit()")
+        public Object evalBindingAddsVarsCached(
+                DynamicObject source,
+                DynamicObject binding,
+                NotProvided file,
+                NotProvided line,
+                @Cached("privatizeRope(source)") Rope cachedSource,
+                @Cached("getBindingDescriptor(binding)") FrameDescriptor bindingDescriptor,
+                @Cached("compileSource(source, getBindingFrame(binding))") RootNodeWrapper cachedRootNode,
+                @Cached("newBindingDescriptor(getContext(), cachedRootNode)") FrameDescriptor newBindingDescriptor,
+                @Cached("compileSource(source, getBindingFrame(binding), newBindingDescriptor)") RootNodeWrapper rootNodeToEval,
+                @Cached("createCallTarget(rootNodeToEval)") CallTarget cachedCallTarget,
+                @Cached("create(cachedCallTarget)") DirectCallNode callNode,
+                @Cached("create()") RopeNodes.EqualNode equalNode) {
+            final MaterializedFrame parentFrame = BindingNodes.newExtrasFrame(binding,
+                    newBindingDescriptor);
+            return eval(rootNodeToEval, cachedCallTarget, callNode, parentFrame);
+        }
+
+        @Specialization(guards = {
+                "isRubyString(source)",
```

Comment:
This EvalNode is becoming huge, we should simplify, by e.g. using defaults and maybe make it a primitive and do some handling in Ruby. But that's for another PR as it would make it harder to review this one.